bootstrap/controller/openshiftbootstrapconfig_controller.go

```diff
@@ -21,21 +21,57 @@ import (
 	"fmt"
 
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 
+	"github.com/go-logr/logr"
 	openshiftclusterv1 "github.com/openshift/api/cluster/v1alpha1"
+	"github.com/openshift/cluster-api-provider-openshift/bootstrap/internal/locking"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	bsutil "sigs.k8s.io/cluster-api/bootstrap/util"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
 )
 
+// InitLocker is a lock that is used around OpenShift cluster initialization.
+type InitLocker interface {
+	Lock(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) bool
+	Unlock(ctx context.Context, cluster *clusterv1.Cluster) bool
+}
+
 // OpenShiftBootstrapConfigReconciler reconciles a OpenShiftBootstrapConfig object.
 type OpenShiftBootstrapConfigReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
+	Client              client.Client
+	SecretCachingClient client.Client
+	Scheme              *runtime.Scheme
+	InitLock            InitLocker
+}
+
+// Scope is a scoped struct used during reconciliation.
+type Scope struct {
+	logr.Logger
+	Config      *openshiftclusterv1.OpenShiftBootstrapConfig
+	ConfigOwner *bsutil.ConfigOwner
+	Cluster     *clusterv1.Cluster
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *OpenShiftBootstrapConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if r.InitLock == nil {
+		r.InitLock = locking.NewControlPlaneInitMutex(mgr.GetClient())
+	}
```

Comment:
For the purpose of this controller of Cluster API, we might still need one to make sure for each cluster we only ever create a single bootstrap data node, and lock the rest of the joins until then.

Machines can be created in parallel, which would lead to multiple bootstrap data being created. The same would apply if there are multiple controllers?

Reply:
Ack, lets keep for now. 

I had imagined with the separate controller, you could set the parallelism of the controller to 1, minor for now so lets keep as is